File: Ruby/Napakalaki/lib/SpecificBadConsequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'BadConsequence.rb'

class SpecificBadConsequence < BadConsequence
  def initialize(p_text, p_levels, tVisible, tHidden)
    super(p_text, p_levels)
    @specificHiddenTreasures= tHidden
    @specificVisibleTreasures= tVisible
  end
  
  attr_reader :specificHiddenTreasures, :specificVisibleTreasures
  
  def adjustToFitTreasureLists(v, h)
    if(!isEmpty)
      sp_v = Array.new
      sp_h = Array.new
      insertado=false
      
      @specificVisibleTreasures.each do |spvi|
        insertado= false
        v.each do |spvp|
          if(!insertado)
            if(spvi==spvp.type)
              if(!sp_v.contains(spvi))
                sp_v << spvi
                insertado=true
              end
            end
          end
        end
      end

      @specificHiddenTreasures.each do |sphi|
        insertado = false
        h.each do |sphp|
          if(!insertado)
            if(sphi==sphp.type)
              if(!sp_h.include?(sphi))
                sp_h << sphi
                insertado = true
              end
            end
          end
        end
      end
      bad_ret = SpecificBadConsequence.new(@text, @levels, sp_v, sp_h);
      return bad_ret
    end
    return self
  end
  
  def substractVisibleTreasure(t)
    if(!@specificVisibleTreasures.empty?)
      @specificVisibleTreasures.delete(t.type)
    end
  end
  
  def substractHiddenTreasure(t)
    if(!@specificHiddenTreasures.empty?)
      @specificHiddenTreasures.delete(t.type)
    end
  end
  
  def isEmpty
    return @specificHiddenTreasures.empty? && @specificVisibleTreasures.empty?
  end
  
  def to_s
    "\tDescripcion: #{@text} \n\tPerderas niveles: #{@levels} \n\tPerderas estos tesoros especificos visibles: #{@specificVisibleTreasures} \n\tPerderas estos tesoros especificos ocultos: #{@specificHiddenTreasures}"
  end
end

File: Java/Napakalaki/src/napakalaki/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;
import java.util.ArrayList;
/**
 *
 * @author juane
 */
public class SpecificBadConsequence extends BadConsequence{
    private ArrayList<TreasureKind> specificHiddenTreasures= new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures= new ArrayList();
    
    public SpecificBadConsequence(String p_text, int p_levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden){
        super(p_text, p_levels);
        this.specificHiddenTreasures= tHidden;
        this.specificVisibleTreasures= tVisible;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
        BadConsequence bad_ret;
        
        if(!isEmpty()){
                boolean insertado=false;
                
                ArrayList<TreasureKind> sp_v=new ArrayList(), sp_h=new ArrayList();
                
                for(TreasureKind spvi :specificVisibleTreasures){
                    insertado= false;
                    for(Treasure spvp :v){
                        if(!insertado)
                            if(spvi==spvp.getType()){
                                if(!sp_v.contains(spvi)){
                                    sp_v.add(spvi);
                                    insertado=true;
                                }
                            }
                    }
                }
                
                for(TreasureKind sphi :specificHiddenTreasures){
                    insertado= false;
                    for(Treasure sphp :h){
                        if(!insertado)
                            if(sphi==sphp.getType()){
                                if(!sp_h.contains(sphi)){
                                    sp_h.add(sphi);
                                    insertado= true;
                                }
                            }
                    }
                }
                bad_ret= new SpecificBadConsequence(text, levels, sp_v, sp_h);
            
            return bad_ret;
        }
        return this;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
            if(!specificVisibleTreasures.isEmpty())
                specificVisibleTreasures.remove(t.getType());
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
            if(!specificHiddenTreasures.isEmpty())
                specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public Boolean isEmpty(){
        return specificHiddenTreasures.isEmpty() && specificVisibleTreasures.isEmpty();
    }
    
    @Override
    public String toString(){
        return "\n\nDescripcion: " +text + 
                "\nPerderas niveles: "+ Integer.toString(levels)+
                "\nPerderas estos tesoros especificos visibles: " + specificVisibleTreasures.toString() +
                "\nPerderas estos tesoros especificos ocultos: " + specificHiddenTreasures.toString();
    }
}
